spec.md

```diff
@@ -2563,46 +2563,87 @@ Compiler _implementation_.  A FIRRTL Compiler is a program that converts FIRRTL
 text to another representation, e.g., Verilog, VHDL, a programming language, or
 a binary program.
 
-## Aggregate Type Lowering (Lower Types)
+## Module Conventions
 
-A FIRRTL Compiler should provide a "Lower Types" pass that converts aggregate
-types to ground types.
+A module's convention describes how it's ports are lowered to the output format,
+and serves as a kind of ABI for modules.
 
-A FIRRTL Compiler must apply such a pass to the ports of all "public" modules in
-a Verilog/VHDL representation.  Public modules are defined as (1) the top-level
-module and (2) any external modules.
+### The "Scalarized" Convention
 
-A FIRRTL Compiler may apply such a pass to other types in a FIRRTL circuit.
+The scalarized convention lowers aggregate ports to ground values. The
+scalarized convention should be the default convention for "public" modules,
+such as the top module of a circuit, or a "device under test".
 
 The Lower Types algorithm operates as follows:
 
-1. Ground type names are unmodified.
+1. Ports are scalarized top-to-bottom.
 
-2. Vector types are converted to ground types by appending a suffix, `_<i>`, to
-   the i^th^ element of the vector.  (`<` and `>` are not included in the
-   suffix.)
+2. Ground-typed ports names are unmodified.
 
-3. Bundle types are converted to ground types by appending a suffix, `_<name>`,
-   to the field called `name`.  (`<` and `>` are not included in the suffix.)
+3. Vector-typed ports are scalarized to ground-typed ports by appending a
+   suffix, `_<i>`, to the i^th^ element of the vector. Elements are scalarized
+   recursively, depth-first, and left-to-right.
 
-New names generated by Lower Types must be unique with respect to the current
-namespace (see [@sec:namespaces]).
+4. Bundle-typed ports are scalarized to ground-typed ports by appending a
+   suffix, `_<name>`, to the field called `name`. fields are scalarized
+   recursively, depth-first, and left-to-right.
 
-E.g., consider the following wire:
+E.g., consider the following port:
 
 ``` firrtl
-wire a : { b: UInt<1>, c: UInt<2> }[2]
+module Top :
+  input a : { b: UInt<1>, c: UInt<2> }[2]
 ```
 
-The result of a Lower Types pass applied to this wire is:
+Scalarization breaks `a` into the following ports:
 
 ``` firrtl
-wire a_0_b : UInt<1>
-wire a_0_c : UInt<2>
-wire a_1_b : UInt<1>
-wire a_1_c : UInt<2>
+module Top :
+  input a_0_b : UInt<1>  ;; a[0].b
+  input a_0_c : UInt<2>  ;; a[0].c
+  input a_1_b : UInt<1>  ;; a[1].b
+  input a_1_c : UInt<2>  ;; a[1].c
 ```
 
+New names generated by port scalarization must be unique with respect to the
+current namespace (see [@sec:namespaces]). In the case of a collision after
+renaming, priority will be given to values that are scalarized first.
```

Comment:
I don't think that this is exactly the right description given how the lowering algorithm is described. If this is depth first and incremental, then there are points where the namespace guarantees are violated if the namespace is pre-populated.

I'd add some language that "a new, empty namespace is created for lowering ports". "As new names are added, names must be unique with respect to this namespace."